modules/light-clients/08-wasm/types/querier.go

```diff
@@ -0,0 +1,150 @@
+
+	errorsmod "cosmossdk.io/errors"
+	storetypes "cosmossdk.io/store/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	abci "github.com/cometbft/cometbft/abci/types"
+
+	"github.com/cosmos/ibc-go/modules/light-clients/08-wasm/internal/ibcwasm"
+)
+
+var (
+	_ wasmvmtypes.Querier = (*DefaultQuerier)(nil)
+
+	queryPlugins = NewDefaultQueryPlugins()
+)
+
+type DefaultQuerier struct {
+	Ctx      sdk.Context
+	CallerID string
+}
+
+// NewDefaultQuerier returns a default querier that can be used in the contract.
+func NewQueryHandler(ctx sdk.Context, callerID string) *DefaultQuerier {
+	return &DefaultQuerier{
+		Ctx:      ctx,
+		CallerID: callerID,
+	}
+}
+
+// GasConsumed implements the wasmvmtypes.Querier interface.
+func (q *DefaultQuerier) GasConsumed() uint64 {
+	return VMGasRegister.ToWasmVMGas(q.Ctx.GasMeter().GasConsumed())
+}
+
+// Query implements the wasmvmtypes.Querier interface.
+func (q *DefaultQuerier) Query(request wasmvmtypes.QueryRequest, gasLimit uint64) ([]byte, error) {
+	sdkGas := VMGasRegister.FromWasmVMGas(gasLimit)
+
+	subCtx, _ := q.Ctx.WithGasMeter(storetypes.NewGasMeter(sdkGas)).CacheContext()
+
+	// make sure we charge the higher level context even on panic
+	defer func() {
+		q.Ctx.GasMeter().ConsumeGas(subCtx.GasMeter().GasConsumed(), "contract sub-query")
+	}()
+
+	if request.Stargate != nil {
+		return GetQueryPlugins().Stargate(subCtx, request.Stargate)
+	}
+
+	if request.Custom != nil {
+		return GetQueryPlugins().Custom(subCtx, request.Custom)
+	}
+
+	return nil, wasmvmtypes.UnsupportedRequest{Kind: "Unsupported query request"}
+}
+
+type (
+	CustomQuerier   func(ctx sdk.Context, request json.RawMessage) ([]byte, error)
+	StargateQuerier func(ctx sdk.Context, request *wasmvmtypes.StargateQuery) ([]byte, error)
+)
+
+// QueryPlugins is a list of query handlers that can be used to extend the default querier.
+type QueryPlugins struct {
+	Custom   CustomQuerier
+	Stargate StargateQuerier
+}
+
+// Merge merges the query plugin with a provided one.
+func (e QueryPlugins) Merge(x *QueryPlugins) QueryPlugins {
+	// only update if this is non-nil and then only set values
+	if x == nil {
+		return e
+	}
+
+	if x.Custom != nil {
+		e.Custom = x.Custom
+	}
+
+	if x.Stargate != nil {
+		e.Stargate = x.Stargate
+	}
+
+	return e
+}
+
+// SetQueryPlugins sets the current query plugins
+func SetQueryPlugins(plugins *QueryPlugins) {
+	queryPlugins = plugins
+}
+
+// GetQueryPlugins returns the current query plugins
+func GetQueryPlugins() *QueryPlugins {
+	return queryPlugins
+}
+
+// NewDefaultQueryPlugins returns the default set of query plugins
+func NewDefaultQueryPlugins() *QueryPlugins {
+	return &QueryPlugins{
+		Custom:   RejectCustomQuerier(),
+		Stargate: AcceptListStargateQuerier([]string{}),
+	}
+}
+
+// AcceptListStargateQuerier allows all queries that are in the accept list provided and in the default accept list.
+// This function returns protobuf encoded responses.
```

Comment:
maybe mention its bytes as opposed to a proto.Message

```suggestion
// This function returns protobuf encoded responses in bytes.
```